libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/verification/RustSessionVerificationService.kt

```diff
@@ -165,8 +181,27 @@ class RustSessionVerificationService(
 
     override fun didFinish() {
         _verificationFlowState.value = VerificationFlowState.Finished
-        // Ideally this should be `verificationController?.isVerified().orFalse()` but for some reason it always returns false
-        updateVerificationStatus(VerificationState.VERIFIED)
+        val sessionId = client.userId()
+        sessionCoroutineScope.launch {
+            // Ideally this should be `verificationController?.isVerified().orFalse()` but for some reason it returns false if run immediately
+            // It also sometimes unexpectedly fails to report the session as verified, so we have to handle that possibility and fail if needed
+            runCatching {
+                withTimeout(5.seconds) {
+                    while (!verificationController.isVerified()) {
+                        delay(100)
+                    }
+                }
+            }
+                .onSuccess {
+                    val existingSession = sessionStore.getSession(sessionId) ?: error("No session to restore with id $sessionId")
+                    sessionStore.updateData(existingSession.copy(needsVerification = false))
```

Comment:
I see that setting `needsVerification` to `false` is done here, so I guess my first comment should just be "just remove this code"?

Reply:
Or also set the boolean in case of failure?